File: sdks/python/speechswitch/providers/mistral.py
"""Mistral Voxtral's documented protocol; its partial OpenAPI is not codegen input."""

import asyncio
import base64
import json
import math
import os
import re
from collections.abc import AsyncGenerator, AsyncIterator, Mapping
from contextlib import aclosing, asynccontextmanager
from typing import NoReturn, cast
from urllib.parse import urlsplit, urlunsplit

from speechswitch.generated.auth import Auth
from speechswitch.generated.mistral import TtsRequest
from speechswitch.generated.mistral_output import PromptTokensDetails, PromptTokensDetailsMessagesItem, SynthesisItem, Usage
from speechswitch.generated.validators.mistral import REQUEST_DEFAULTS, validate_request
from speechswitch.http import AudioStream, HttpRequest, HttpResponse, HttpTransport
from speechswitch.sse import SseDecoder
from speechswitch.validation import is_mapping, is_sequence


class MistralError(Exception):
    def __init__(self, status_code: int, body: str, retry_after: str | None) -> None:
        self.status_code = status_code
        self.body = body
        self.retry_after = retry_after
        super().__init__(f"Mistral synthesis failed ({status_code})")


def _object(value: object) -> Mapping[object, object]:
    if not is_mapping(value):
        raise TypeError("Mistral returned an invalid response object")
    return value


def _count(value: object, field: str) -> int:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not 0 <= value <= 9007199254740991 or not math.isfinite(value) or value % 1 != 0:
        raise TypeError(f"Mistral returned invalid {field}")
    return int(value)


def _audio_data(value: object) -> bytes:
    if not isinstance(value, str) or re.fullmatch(r"(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?", value) is None:
        raise TypeError("Mistral returned invalid base64 audio")
    return base64.b64decode(value, validate=True)


def _usage_message(value: object) -> PromptTokensDetailsMessagesItem:
    message = _object(value)
    role = message.get("role")
    if role != "system" and role != "user" and role != "assistant" and role != "tool":
        raise TypeError("Mistral returned invalid usage message role")
    truncated = message.get("truncated")
    if "truncated" in message and not isinstance(truncated, bool):
        raise TypeError("Mistral returned invalid usage truncated flag")
    return cast(PromptTokensDetailsMessagesItem, {
        "role": role,
        **({"total_tokens": None if message["total_tokens"] is None else _count(message["total_tokens"], "total_tokens")} if "total_tokens" in message else {}),
        **({"truncated": truncated} if isinstance(truncated, bool) else {}),
        **({"usage_count": _count(message["usage_count"], "usage_count")} if "usage_count" in message else {}),
    })


def _prompt_details(value: object) -> PromptTokensDetails:
    details = _object(value)
    messages = details.get("messages")
    if "messages" in details and not is_sequence(messages):
        raise TypeError("Mistral returned invalid usage messages")
    return cast(PromptTokensDetails, {
        **({"cached_tokens": _count(details["cached_tokens"], "cached_tokens")} if "cached_tokens" in details else {}),
        **({"audio_tokens": _count(details["audio_tokens"], "audio_tokens")} if "audio_tokens" in details else {}),
        **({"messages": [_usage_message(message) for message in messages]} if is_sequence(messages) else {}),
    })


def _usage(value: object) -> Usage:
    usage = _object(value)
    completion = usage.get("completion_tokens_details")
    details = None if completion is None else _object(completion)
    return cast(Usage, {
        **({"prompt_tokens": _count(usage["prompt_tokens"], "prompt_tokens")} if "prompt_tokens" in usage else {}),
        **({"completion_tokens": None if usage["completion_tokens"] is None else _count(usage["completion_tokens"], "completion_tokens")} if "completion_tokens" in usage else {}),
        **({"total_tokens": _count(usage["total_tokens"], "total_tokens")} if "total_tokens" in usage else {}),
        **({"prompt_audio_seconds": None if usage["prompt_audio_seconds"] is None else _count(usage["prompt_audio_seconds"], "prompt_audio_seconds")} if "prompt_audio_seconds" in usage else {}),
        **({"request_count": None if usage["request_count"] is None else _count(usage["request_count"], "request_count")} if "request_count" in usage else {}),
        **({"cached_tokens": None if usage["num_cached_tokens"] is None else _count(usage["num_cached_tokens"], "num_cached_tokens")} if "num_cached_tokens" in usage else {}),
        **({"prompt_tokens_details": None if usage["prompt_tokens_details"] is None else _prompt_details(usage["prompt_tokens_details"])} if "prompt_tokens_details" in usage else {}),
        **({"prompt_token_details": None if usage["prompt_token_details"] is None else _prompt_details(usage["prompt_token_details"])} if "prompt_token_details" in usage else {}),
        **({"completion_tokens_details": None if details is None else {**({"reasoning_tokens": _count(details["reasoning_tokens"], "reasoning_tokens")} if "reasoning_tokens" in details else {})}} if "completion_tokens_details" in usage else {}),
    })


def _invalid_constant(value: str) -> NoReturn:
    raise TypeError("Mistral returned invalid JSON")


async def _items(response: HttpResponse, body: AudioStream, max_event_bytes: int, max_json_bytes: int) -> AsyncGenerator[SynthesisItem, None]:
    headers = {name.lower(): value for name, value in response.headers.items()}
    content_type = headers.get("content-type", "").split(";", 1)[0].strip().lower()
    if not 200 <= response.status < 300 or content_type == "application/json":
        data = bytearray()
        async for chunk in body:
            if len(data) + len(chunk) > max_json_bytes:
                raise TypeError("Mistral response exceeds max_json_bytes")
            data.extend(chunk)
        text = data.decode("utf-8", errors="replace")
        if not 200 <= response.status < 300:
            raise MistralError(response.status, text, headers.get("retry-after"))
        value: object = json.loads(text, parse_constant=_invalid_constant)
        audio = _audio_data(_object(value).get("audio_data"))
        if not audio:
            raise TypeError("Mistral returned no audio")
        yield audio
        yield {"event": "done"}
        return
    if content_type != "text/event-stream":
        raise TypeError("Mistral returned an unsupported response content type")
    decoder = SseDecoder(max_event_bytes)
    received_audio = False
    try:
        async for chunk in body:
            for byte in chunk:
                message = decoder.push(byte)
                if message is None:
                    continue
                value: object = json.loads(message["data"], parse_constant=_invalid_constant)
                packet = _object(value)
                kind = packet.get("type")
                if kind is None:
                    kind = message["event"]
                if message["event"] != "message" and message["event"] != kind:
                    raise TypeError("Mistral returned conflicting SSE event types")
                if kind == "speech.audio.delta":
                    audio = _audio_data(packet.get("audio_data"))
                    if audio:
                        received_audio = True
                        yield audio
                elif kind == "speech.audio.done":
                    usage = _usage(packet.get("usage"))
                    if not received_audio:
                        raise TypeError("Mistral returned no audio")
                    await body.aclose()
                    yield {"event": "done", "usage": usage}
                    return
                else:
                    raise TypeError("Mistral returned an unsupported speech event")
        raise TypeError("Mistral speech stream ended before speech.audio.done")
    finally:
        decoder.finish()


def _json_container(value: object) -> object:
    if is_mapping(value):
        return dict(value)
    if is_sequence(value):
        return list(value)
    raise TypeError("Mistral request contains unsupported JSON data")


@asynccontextmanager
async def synthesize(request: TtsRequest, *, transport: HttpTransport, auth: Auth | None = None,
                     base_url: str = "https://api.mistral.ai", timeout_ms: int | None = None,
                     max_event_bytes: int = 4 * 1024 * 1024, max_json_bytes: int = 16 * 1024 * 1024) -> AsyncIterator[AsyncIterator[SynthesisItem]]:
    """Use async with for deterministic cleanup, including early loop exit.

    The injected transport must return at headers, honor task cancellation and
    never redirect credential-bearing requests across origins. Deadlines cover
    the whole context, not just HTTP headers. Audio is always returned as bytes.
    """
    validate_request(request)
    key = (auth or {}).get("mistral", {}).get("api_key")
    if key is None:
        key = os.environ.get("SPEECHSWITCH_MISTRAL_API_KEY", os.environ.get("MISTRAL_API_KEY"))
    if not key:
        raise TypeError("Missing auth.mistral.apiKey configuration")
    if timeout_ms is not None and (type(timeout_ms) is not int or not 0 <= timeout_ms <= 2147483647):
        raise TypeError("Mistral timeout_ms must be an integer between 0 and 2147483647")
    if timeout_ms == 0:
        raise TimeoutError("Mistral synthesis deadline expired")
    for name, limit in [("max_event_bytes", max_event_bytes), ("max_json_bytes", max_json_bytes)]:
        if type(limit) is not int or limit <= 0:
            raise TypeError(f"Mistral {name} must be a positive integer")
    url = urlsplit(base_url)
    if url.scheme not in ("http", "https") or not url.hostname or url.username is not None or url.password is not None or url.fragment:
        raise TypeError("Mistral base_url must be an HTTP(S) URL without credentials or a fragment")
    target = urlunsplit((url.scheme, url.netloc, url.path.rstrip("/") + "/v1/audio/speech", url.query, ""))
    output = request.get("output")
    payload: dict[str, object] = {
        "model": request.get("model", REQUEST_DEFAULTS["model"]), "input": request["text"], "stream": True,
        "response_format": output["format"] if output is not None else "pcm",
        **({"voice_id": request["voice"]} if "voice" in request else {}),
        **({"ref_audio": base64.b64encode(request["reference_audio"]).decode("ascii")} if "reference_audio" in request else {}),
        **({"metadata": request["metadata"]} if "metadata" in request else {}),
        **({"prompt_cache_key": request["prompt_cache_key"]} if "prompt_cache_key" in request else {}),
    }
    wire = HttpRequest("POST", target, {"Authorization": f"Bearer {key}", "Content-Type": "application/json", "Accept": "text/event-stream, application/json"}, json.dumps(payload, allow_nan=False, default=_json_container, separators=(",", ":")).encode("utf-8"))
    async with asyncio.timeout(None if timeout_ms is None else timeout_ms / 1000):
        response = await transport.send(wire)
        body = AudioStream(response.body)
        try:
            async with aclosing(_items(response, body, max_event_bytes, max_json_bytes)) as items:
                yield items
        except BaseException:
            try:
                await body.aclose()
            except Exception:
                pass
            raise
        finally:
            await body.aclose()

File: sdks/go/providers/mistral/mistral.go
// Package mistral implements the documented Voxtral protocol directly. The
// cataloged OpenAPI contracts are incomplete and are not generation sources.
package mistral

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/speechswitch/client/sdks/go/generated/auth"
	schema "github.com/speechswitch/client/sdks/go/generated/mistral"
	out "github.com/speechswitch/client/sdks/go/generated/mistral_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

type Options struct {
	Auth auth.Auth
	// Nil uses net/http with redirects disabled. Overrides must honor cancellation
	// and must not forward credentials across origins on redirects.
	Transport runtime.HTTPTransport
	BaseURL   string
	// Omission has no deadline; an explicit zero expires before network access.
	Timeout runtime.Optional[time.Duration]
	// Zero selects the defaults: 4 MiB per SSE event, 16 MiB per JSON/error body.
	MaxEventBytes int
	MaxJSONBytes  int
}

// Synthesize always returns a pull-based stream. Defer Close after opening it.
// Context/deadline cancellation releases the response even between Next calls.
func Synthesize(ctx context.Context, request schema.TtsRequest, options Options) (runtime.Input[out.SynthesisItem], error) {
	if _, err := schema.ValidateRequest(request); err != nil {
		return nil, err
	}
	var key string
	if options.Auth.Mistral.Present && options.Auth.Mistral.Value.ApiKey.Present {
		key = options.Auth.Mistral.Value.ApiKey.Value
	} else if scoped, present := os.LookupEnv("SPEECHSWITCH_MISTRAL_API_KEY"); present {
		key = scoped
	} else {
		key = os.Getenv("MISTRAL_API_KEY")
	}
	if key == "" {
		return nil, errors.New("Missing auth.mistral.apiKey configuration")
	}
	if options.Timeout.Present && options.Timeout.Value < 0 {
		return nil, errors.New("Mistral Timeout must not be negative")
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if options.Timeout.Present && options.Timeout.Value == 0 {
		return nil, context.DeadlineExceeded
	}
	maxEventBytes, maxJSONBytes := options.MaxEventBytes, options.MaxJSONBytes
	if maxEventBytes == 0 {
		maxEventBytes = 4 * 1024 * 1024
	}
	if maxJSONBytes == 0 {
		maxJSONBytes = 16 * 1024 * 1024
	}
	if maxEventBytes < 0 || maxJSONBytes < 0 {
		return nil, errors.New("Mistral response byte limits must be positive")
	}
	baseURL := options.BaseURL
	if baseURL == "" {
		baseURL = "https://api.mistral.ai"
	}
	target, err := url.Parse(baseURL)
	if err != nil || target.Hostname() == "" || target.User != nil || target.Fragment != "" || (target.Scheme != "https" && target.Scheme != "http") {
		return nil, errors.New("Mistral BaseURL must be an HTTP(S) URL without credentials or a fragment")
	}
	escapedPath := strings.TrimRight(target.EscapedPath(), "/") + "/v1/audio/speech"
	target.Path = strings.TrimRight(target.Path, "/") + "/v1/audio/speech"
	target.RawPath = escapedPath
	format := "pcm"
	if request.Output.Present {
		format, err = outputFormat(request.Output.Value)
		if err != nil {
			return nil, err
		}
	}
	payload := map[string]any{"model": request.Model.Value.Value(), "input": request.Text, "stream": true, "response_format": format}
	if request.Voice.Present {
		payload["voice_id"] = request.Voice.Value
	}
	if request.ReferenceAudio.Present {
		payload["ref_audio"] = base64.StdEncoding.EncodeToString(request.ReferenceAudio.Value)
	}
	if request.PromptCacheKey.Present {
		payload["prompt_cache_key"] = request.PromptCacheKey.Value
	}
	if request.Metadata.Present {
		payload["metadata"], err = jsonValue(runtime.JsonObject(request.Metadata.Value))
		if err != nil {
			return nil, err
		}
	}
	encoded, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	var cancel context.CancelFunc
	if options.Timeout.Present {
		ctx, cancel = context.WithTimeout(ctx, options.Timeout.Value)
	} else {
		ctx, cancel = context.WithCancel(ctx)
	}
	wire, err := http.NewRequestWithContext(ctx, http.MethodPost, target.String(), bytes.NewReader(encoded))
	if err != nil {
		cancel()
		return nil, err
	}
	wire.Header.Set("Authorization", "Bearer "+key)
	wire.Header.Set("Content-Type", "application/json")
	wire.Header.Set("Accept", "text/event-stream, application/json")
	transport := options.Transport
	if transport == nil {
		transport = &http.Client{CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}
	}
	response, err := runtime.OpenResponse(wire, transport)
	if err != nil {
		cancel()
		return nil, err
	}
	if err := ctx.Err(); err != nil {
		cancel()
		response.Body.Close()
		return nil, err
	}
	decoder, err := runtime.NewSSEDecoder(maxEventBytes)
	if err != nil {
		cancel()
		response.Body.Close()
		return nil, err
	}
	return &stream{ctx: ctx, cancel: cancel, response: response, decoder: decoder, maxJSONBytes: maxJSONBytes,
		contentType: strings.ToLower(strings.TrimSpace(strings.SplitN(response.Header.Get("Content-Type"), ";", 2)[0]))}, nil
}

type stream struct {
	ctx           context.Context
	cancel        context.CancelFunc
	response      *runtime.HTTPResponse
	decoder       *runtime.SSEDecoder
	contentType   string
	maxJSONBytes  int
	once          sync.Once
	closeError    error
	mutex         sync.Mutex
	buffer        []byte
	receivedAudio bool
	terminal      bool
	jsonDone      bool
}

func (s *stream) closeBody() error {
	s.once.Do(func() { s.cancel(); s.closeError = s.response.Body.Close() })
	return s.closeError
}

func (s *stream) Close() error {
	err := s.closeBody()
	s.mutex.Lock()
	s.buffer = nil
	s.decoder.Finish()
	s.mutex.Unlock()
	return err
}

func (s *stream) Next(ctx context.Context) (out.SynthesisItem, error) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal {
		return nil, io.EOF
	}
	item, err := s.next(ctx)
	if err == nil {
		err = ctx.Err()
		if err == nil {
			err = s.ctx.Err()
		}
		if err != nil {
			item = nil
		}
	}
	_, done := item.(out.SynthesisItemAsDone)
	if err != nil || done {
		s.terminal = true
		s.closeBody()
		s.buffer = nil
		s.decoder.Finish()
	}
	return item, err
}

func (s *stream) next(ctx context.Context) (out.SynthesisItem, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := s.ctx.Err(); err != nil {
		return nil, err
	}
	if s.jsonDone {
		return out.SynthesisItemAsDone{}, nil
	}
	if s.response.StatusCode < 200 || s.response.StatusCode >= 300 || s.contentType == "application/json" {
		var data []byte
		for {
			chunk, err := s.response.Body.Next(ctx)
			if err != nil && err != io.EOF {
				return nil, err
			}
			if len(chunk) > s.maxJSONBytes-len(data) {
				return nil, errors.New("Mistral response exceeds MaxJSONBytes")
			}
			data = append(data, chunk...)
			if err == io.EOF {
				break
			}
		}
		if s.response.StatusCode < 200 || s.response.StatusCode >= 300 {
			retry := runtime.Optional[string]{}
			if values, present := s.response.Header["Retry-After"]; present && len(values) > 0 {
				retry = runtime.Some(values[0])
			}
			return nil, &Error{StatusCode: s.response.StatusCode, Body: runtime.DecodeUTF8(data), RetryAfter: retry}
		}
		value, err := object(data)
		if err != nil {
			return nil, err
		}
		audio, err := audioData(value["audio_data"])
		if err != nil {
			return nil, err
		}
		if len(audio) == 0 {
			return nil, errors.New("Mistral returned no audio")
		}
		s.jsonDone = true
		return out.SynthesisItemAsBytes{Value: audio}, nil
	}
	if s.contentType != "text/event-stream" {
		return nil, errors.New("Mistral returned an unsupported response content type")
	}
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if err := s.ctx.Err(); err != nil {
			return nil, err
		}
		if len(s.buffer) == 0 {
			chunk, err := s.response.Body.Next(ctx)
			if err == io.EOF {
				return nil, errors.New("Mistral speech stream ended before speech.audio.done")
			}
			if err != nil {
				return nil, err
			}
			s.buffer = chunk
		}
		for len(s.buffer) > 0 {
			message, err := s.decoder.Push(s.buffer[0])
			s.buffer = s.buffer[1:]
			if err != nil {
				return nil, err
			}
			if message == nil {
				continue
			}
			item, err := decodeEvent(*message)
			if err != nil {
				return nil, err
			}
			if audio, ok := item.(out.SynthesisItemAsBytes); ok {
				if len(audio.Value) == 0 {
					continue
				}
				s.receivedAudio = true
			} else if !s.receivedAudio {
				return nil, errors.New("Mistral returned no audio")
			}
			return item, nil
		}
	}
}

func outputFormat(value schema.TtsRequestOutput) (string, error) {
	switch value := value.(type) {
	case schema.TtsRequestOutputAsPcm:
		return value.Value.Format.Value(), nil
	case *schema.TtsRequestOutputAsPcm:
		return value.Value.Format.Value(), nil
	case schema.TtsRequestOutputAsObject:
		return encodedFormat(value.Value.Format)
	case *schema.TtsRequestOutputAsObject:
		return encodedFormat(value.Value.Format)
	default:
		return "", errors.New("Unsupported generated Mistral output representation")
	}
}

func encodedFormat(value schema.TtsRequestOutputObjectFormat) (string, error) {
	switch value := value.(type) {
	case schema.TtsRequestOutputObjectFormatAsFlac:
		return value.Value.Value(), nil
	case *schema.TtsRequestOutputObjectFormatAsFlac:
		return value.Value.Value(), nil
	case schema.TtsRequestOutputObjectFormatAsMp3:
		return value.Value.Value(), nil
	case *schema.TtsRequestOutputObjectFormatAsMp3:
		return value.Value.Value(), nil
	case schema.TtsRequestOutputObjectFormatAsOpus:
		return value.Value.Value(), nil
	case *schema.TtsRequestOutputObjectFormatAsOpus:
		return value.Value.Value(), nil
	case schema.TtsRequestOutputObjectFormatAsWav:
		return value.Value.Value(), nil
	case *schema.TtsRequestOutputObjectFormatAsWav:
		return value.Value.Value(), nil
	default:
		return "", errors.New("Unsupported generated Mistral format representation")
	}
}

// Called only after generated validation: convert the closed JSON algebra to
// encoding/json values, including nil collections as empty rather than null.
func jsonValue(value runtime.JsonValue) (any, error) {
	switch value := value.(type) {
	case runtime.JsonNull, *runtime.JsonNull:
		return nil, nil
	case runtime.JsonBool:
		return bool(value), nil
	case *runtime.JsonBool:
		return bool(*value), nil
	case runtime.JsonNumber:
		return float64(value), nil
	case *runtime.JsonNumber:
		return float64(*value), nil
	case runtime.JsonString:
		return string(value), nil
	case *runtime.JsonString:
		return string(*value), nil
	case *runtime.JsonArray:
		return jsonValue(*value)
	case *runtime.JsonObject:
		return jsonValue(*value)
	case runtime.JsonArray:
		result := make([]any, len(value))
		for index, child := range value {
			converted, err := jsonValue(child)
			if err != nil {
				return nil, err
			}
			result[index] = converted
		}
		return result, nil
	case runtime.JsonObject:
		result := make(map[string]any, len(value))
		for key, child := range value {
			converted, err := jsonValue(child)
			if err != nil {
				return nil, err
			}
			result[key] = converted
		}
		return result, nil
	default:
		return nil, errors.New("Unsupported generated JSON representation")
	}
}
